### ISSUE TITLE
False positive error when using a NewType as a defaultdict's default_factory

### ISSUE THREAD

**Issue by jonasrauber** [python/mypy#14339]
**Actual Behavior**
```python
from typing import NewType
from collections import defaultdict

UserId = NewType("UserId", int)

defaultdict[str, UserId](UserId)  # mypy error
```

Gives the following false-positive error:
`error: Argument 1 to "defaultdict" has incompatible type "Type[UserId]"; expected "Optional[Callable[[], UserId]]"  [arg-type]`

**Expected Behavior**

No error, like a simple subclass:

```python
from typing import NewType
from collections import defaultdict

class UserId(int):
    pass

defaultdict[str, UserId](UserId)  # no error
```

**Comment by jonasrauber** [comment]
FWIW, the following workaround works (of course):

```python
from typing import NewType
from collections import defaultdict

UserId = NewType("UserId", int)

defaultdict[str, UserId](lambda: UserId(0))
```

**Comment by erictraut** [comment]
Mypy is correct here because `UserId` cannot be called with zero arguments. Your first code sample will therefore crash when defaultdict attempts to allocate a new default value.

```python
UserId() # TypeError: _typing._idfunc() takes exactly one argument (0 given)
```

The correct way to handle this is to provide a lambda, as you've done in your second example.
